### PR TITLE
feat: compile distroless python3 OOM container

### PR DESCRIPTION
**Pull request by KrawczowaKris**
# Summary
## Test Task
Compile container using distroless base python3 image https://github.com/GoogleContainerTools/distroless

See python files in testtask folder.

What you will need to do:
-  setup pre-commit in the folder to run https://github.com/hadolint/hadolint checks against any Dockerfiles
        don't open any PR if your Dockerfile does not pass validation from the linter.
-  provide README.md with instructions how to run/test your work
-  create Makefile with the following targets:
        - `build`
        - `run`
            `run` target should accept parameter `limit` that will set the desired max memory limit for the running docker container
            if such parameter is not set, a default value of 100m should be used
-  make run should start the docker container and output content similar to
```
    Usage over time: [20.67578125, 20.67578125, 35.30078125, 50.17578125, 64.80078125, 79.55078125, 94.30078125, 20.734375]
    Peak usage: 94.30078125
```
-  make run with optional argument for memory set at 30m should understand that container exited by error and in this case output the following line
```
    {exited false false false true false 0 137  2023-04-17T00:50:11.053754525Z 2023-04-17T00:50:13.721847306Z <nil>}
```
which is content of the docker State value.

## my results
`make run`

**output:**
```
Usage over time: [22.12890625, 22.16796875, 36.51171875, 50.83984375, 65.453125, 79.890625, 94.328125, 22.2890625]
Peak usage: 94.328125
```
`make run MEMORY_LIMIT=110M `

**output:**
```
Usage over time: [20.09765625, 20.16796875, 37.58203125, 54.81640625, 72.41796875, 90.20703125, 20.23046875]
Peak usage: 90.20703125
```
`make run MEMORY_LIMIT=30M`
**output:**
```
{exited false false false true false 0 137  2023-04-25T10:53:29.528821529Z 2023-04-25T10:53:29.967758705Z <nil>}
```

### CODE REVIEW

File: chapter4/compile-distroless/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+FROM python:3-slim AS build-env
+COPY testtask /memory
+WORKDIR /memory
+RUN pip install --no-cache-dir -r requirements.txt
+
+# Deploy
+FROM gcr.io/distroless/python3:debug
+WORKDIR /memory
+COPY --from=build-env /memory .
+COPY --from=build-env /usr/local /usr/local
```
Comment (dmitry-mightydevops) [inline chapter4/compile-distroless/Dockerfile:11]: that is not safe - only copy what is python related files changes on line 5 above.

File: chapter4/compile-distroless/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,12 @@
+# Build
+FROM python:3-slim AS build-env
+COPY testtask /memory
+WORKDIR /memory
+RUN pip install --no-cache-dir -r requirements.txt
+
+# Deploy
+FROM gcr.io/distroless/python3:debug
```
Comment (dmitry-mightydevops) [inline chapter4/compile-distroless/Dockerfile:8]: don't use debug for prod

File: chapter4/compile-distroless/Makefile [hunk 3]
```diff
@@ -0,0 +1,6 @@
+build:
+	docker build . -t memory-test
+
+MEMORY_LIMIT=100M
+run:
```
Comment (dmitry-mightydevops) [inline chapter4/compile-distroless/Makefile:5]: 1. your `run` should work even without MEMORY_LIMIT passed.
2. extract `memory-container` value into makefile constant.

**Review by dmitry-mightydevops — CHANGES_REQUESTED** [review]

File: chapter4/compile-distroless/Dockerfile (hunk 1 above)
Comment (KrawczowaKris) [reply chapter4/compile-distroless/Dockerfile:11]: remade

File: chapter4/compile-distroless/Makefile (hunk 3 above)
Comment (KrawczowaKris) [reply chapter4/compile-distroless/Makefile:5]: I don't quite understand the first comment

Comment (dmitry-mightydevops) [reply chapter4/compile-distroless/Makefile:6]: correct

**Review by dmitry-mightydevops — APPROVED** [review]

**Pull request merged**